molvis/molvisbase/src/datastructures/molecularstructure.cpp

```diff
@@ -0,0 +1,229 @@
+
+const std::string& MolecularStructure::getSource() const { return source_; }
+
+size_t MolecularStructure::getAtomCount() const { return atoms_.positions.size(); }
+
+size_t MolecularStructure::getResidueCount() const { return residues_.size(); }
+
+size_t MolecularStructure::getChainCount() const { return chains_.size(); }
+
+size_t MolecularStructure::getBondCount() const { return bonds_.size(); }
+
+void MolecularStructure::setAtoms(Atoms atoms) { atoms_ = std::move(atoms); }
+
+void MolecularStructure::setResidues(std::vector<Residue> residues) {
+    residues_ = std::move(residues);
+}
+
+void MolecularStructure::setChains(std::vector<Chain> chains) { chains_ = std::move(chains); }
+
+const Atoms& MolecularStructure::getAtoms() const { return atoms_; }
+
+const std::vector<Residue>& MolecularStructure::getResidues() const { return residues_; }
+
+const std::vector<Chain>& MolecularStructure::getChains() const { return chains_; }
+
+std::optional<Residue> MolecularStructure::getResidue(size_t residueId, size_t chainId) const {
+    auto it = util::find_if(residues_,
+                            [&](auto& r) { return (r.id == residueId) && (r.chainId == chainId); });
+    if (it != residues_.end()) {
+        return *it;
+    } else {
+        return std::nullopt;
+    }
+}
+
+std::optional<Chain> MolecularStructure::getChain(size_t chainId) const {
+    auto it = util::find_if(chains_, [id = chainId](auto& r) { return r.id == id; });
+    if (it != chains_.end()) {
+        return *it;
+    } else {
+        return std::nullopt;
+    }
+}
+
+std::optional<size_t> MolecularStructure::getAtomIndex(const std::string& fullAtomName,
+                                                       size_t residueId, size_t chainId) const {
+    auto residue = getResidue(residueId, chainId);
+    if (residue) {
+        for (auto i : residue->atoms) {
+            if ((atoms_.fullNames[i] == fullAtomName) && (atoms_.residueIds[i] == residueId) &&
+                (atoms_.chainIds[i] == chainId)) {
+                return i;
+            }
+        }
+    } else {
+        for (size_t i = 0; i < atoms_.fullNames.size(); ++i) {
+            if ((atoms_.fullNames[i] == fullAtomName) && (atoms_.residueIds[i] == residueId) &&
+                (atoms_.chainIds[i] == chainId)) {
+                return i;
+            }
+        }
+    }
+    return std::nullopt;
+}
+
+void MolecularStructure::updateStructure() {
+    atoms_.residueIndices.clear();
+
+    for (size_t i = 0; i < atoms_.size(); ++i) {
+        const auto resId = atoms_.residueIds[i];
+        const auto chainId = atoms_.chainIds[i];
+        auto it = util::find_if(residues_, [id = resId](const Residue& r) { return r.id == id; });
+        if (it == residues_.end()) {
+            throw Exception(fmt::format("Invalid residue ID '{}'", resId), IVW_CONTEXT);
+        }
+        IVW_ASSERT(chainId == it->chainId, "Chain ID mismatch");
```

Comment:
This should also throw since it is a runtime error due to incorrect input data?

Reply:
The constructor of `MolecularStructure` will now throw if the input data is invalid.